lib/checkstl.cpp

```diff
@@ -229,10 +237,48 @@ static std::string getContainerName(const Token *containerToken)
     return ret;
 }
 
+static const Token* findIteratorContainer(const Token* start, const Token* end, unsigned int id)
+{
+    const Token* containerToken = nullptr;
+    for(const Token* tok = start; tok != end; tok = tok->next())
+    {
+        if (Token::Match(tok, "%varid% = %name% .", id) &&
+            Token::Match(skipMembers(tok->tokAt(2)), "begin|rbegin|cbegin|crbegin|end|cend|rend|crend|find ("))
```

Comment:
I have the feeling that we can use the `ValueType` instead of this hardcoded pattern. Something like:

    if (Token::Match(tok, "%varid% = %name% . %name% (")) {
        const Token *parTok = tok->tokAt(5);  // (
        if (parTok->valueType() && parTok->valueType()->type == ValueType::Type::ITERATOR)
        ... we have a function call where the result is an iterator ...
    }

Reply:
@danmar Hello! I have replaced the hardcoded pattern.
As for error message, to be honest, I have some doubts about writing message that container is shadowed. In this segment of code we check iterators and, if I'm not mistaken, the situation when one variable is shadowed by another is possible in C+ (it's a warning, yes), something like that:
```
int main()
{
  int a = 10;
  {
    int a = 20;
  }
  return 0;
}
```
But it should be implemented in another part of cppcheck, not in iterators checking. In current variant I write a message to user that iterator is used with different containers but these containers have the same name. Maybe I should improve message in this way?